### ISSUE TITLE
Rebus.Structuremap should have a scanner for IHandleMessages implementations

### ISSUE THREAD

**Issue by jeffreyabecker** [rebus-org/Rebus#471]
Structuremap supports scanning assemblies for types and automagically registering them in the container.  Rebus.Structuremap should contain one which finds non-abstract instances implementing IHandleMessages<> and registers them with an appropriate lifecycle. Something like: [MessageHandlerConvention](https://github.com/jeffreyabecker/Rebus/blob/structuremap-sugar/Rebus.StructureMap/MessageHandlerConvention.cs)

**Comment by mookid8000** [comment]
Does StructureMap pick them up by itself? Or do you tell it somehow to use your `MessageHandlerConvention`?

The assembly with the Windsor container adapter provides a couple of extension methods on `IWindsorContainer` that allow you to go e.g.

    container.AutoRegisterHandlersFromThisAssembly();

or

    container.AutoRegisterHandlersFromAssemblyOf<SomeConcreteMessage>();

Even though Windsor has `IWindsorInstaller`s (which is Windsor's analogue to `IRegistrationConvention`) I felt it was more appropriate to provide extensions directly on the container - this way, one could use them in an installer

    public class RebusHandlersInstaller : IWindsorInstaller
    {
        public void Install(IWindsorContainer container, IConfigurationStore store)
        {
            container.AutoRegisterHandlersFromThisAssembly();
        }
    }

    //.....

    container.Install(From.ThisAssembly());

or just call them directly on the container when initializing it:

    container.Register(
        Classes.FromThisAssembly()
            .BasedOn<ApiController>()
            .WithService..Self()
       )
       .AutoRegisterHandlersFromThisAssembly();

I don't know what the most natural way would be with StructureMap - this is just for inspiration 😁

**Comment by jeffreyabecker** [comment]
With structuremap, you have to have to explicitly add the convention my stupid little test app has a registry that looks like this:
```
public class DefaultRegistry : Registry
{
    public DefaultRegistry()
    {
        Scan(s =>
        {
            s.TheCallingAssembly();
            s.WithDefaultConventions();
            s.WithMessageHanderConvention();
        });
    }
}
```

**Comment by mookid8000** [comment]
What does `WithMessageHanderConvention` contain then?

**Comment by jeffreyabecker** [comment]
WithMessageHandlerConvention is an extension that adds this to the registry:
https://github.com/jeffreyabecker/Rebus/blob/structuremap-sugar/Rebus.StructureMap/MessageHandlerConvention.cs

**Comment by mookid8000** [comment]
oh, it's as simple as that....

I think this would be a neat addition to the SM integration

**Comment by mookid8000** [comment]
fixed by #475 which will be out in 0.99.66
